lib/note-detail.jsx

```diff
@@ -1,62 +1,166 @@
-		if ( this.isValidNote( note ) && this.noteEditor ) {
+		if ( isValidNote( note ) && this.noteEditor ) {
 			this.noteEditor.value = content;
 
 			// Let's focus the editor for new and empty notes
 			if ( content === '' ) {
-				this.noteEditor.focus();
+				invoke( this.noteEditor, 'focus' );
 			}
 		}
-	},
+	};
 
-	componentWillUnmount: function() {
+	componentWillUnmount = () => {
 		window.removeEventListener( 'beforeunload', this.queueNoteSave.flush );
-	},
+		window.removeEventListener( 'keydown', this.interceptTabPresses );
+	};
+
+	handleKey = event => {
+		if (
+			'Enter' !== event.key ||
+			! this.noteEditor ||
+			this.props.previewingMarkdown ||
+			event.shiftKey ||
+			event.altKey ||
+			event.ctrlKey ||
+			event.metaKey
+		) {
+			return true;
+		}
+
+		const {
+			selectionStart,
+			value,
+		} = this.noteEditor;
+
+		const prevLineStart = value.lastIndexOf( '\n', selectionStart - 1 ) + 1;
+		const prevLine = value.substring( prevLineStart, selectionStart - 1 );
+		const prevIndent = prevLine.match( /^(\s+)([-*])/ );
+
+		// The previous line did not start with
+		// indentation, so continue as normal
+		if ( null === prevIndent ) {
+			return true;
+		}
+
+		const indent = prevIndent[ 1 ];
+		const bullet = prevIndent[ 2 ];
+		const nextIndent = `\n${ indent }${ bullet } `;
+
+		// If going a second time in a row that means we
+		// want to escape the list structure, so unwind
+		// and move back to the beginning of the line
+		if ( prevLine.trim() === nextIndent.trim() ) {
+			this.noteEditor.selectionStart = prevLineStart;
+			this.noteEditor.selectionEnd = Math.min( value.indexOf( '\n', prevLineStart ), value.length );
+			document.execCommand( 'delete', false, null );
+			this.noteEditor.selectionStart = prevLineStart - 1;
+			return true;
+		}
+
+		// Prepend the next list prefix to the line
+		document.execCommand( 'insertText', false, nextIndent );
+		this.noteEditor.selectionStart = selectionStart + nextIndent.length;
+		this.noteEditor.selectionEnd = selectionStart + nextIndent.length;
+
+		event.preventDefault();
+		event.stopPropagation();
+
+		return false;
+	};
+
+	interceptTabPresses = event => {
+		if (
+			'Tab' !== event.code ||
+			! this.noteEditor ||
+			'TEXTAREA' !== event.target.nodeName ||
+			this.props.previewingMarkdown ||
+			event.altKey ||
+			event.ctrlKey ||
+			event.metaKey
+		) {
+			return;
+		}
+
+		event.preventDefault();
+		event.stopPropagation();
+
+		const {
+			selectionStart,
+			selectionEnd,
+			value,
+		} = this.noteEditor;
+
+		// if inserting at a cursor position
+		if ( ! event.shiftKey && selectionStart === selectionEnd ) {
+
+			// handle list indents too
+			const currentLineStart = value.lastIndexOf( '\n', selectionStart - 1 ) + 1;
+			const currentLine = value.substring( currentLineStart, selectionStart );
+			console.log( `currentLine: '${ currentLine }'` );
+
+			// line is a list point
+			if ( /^\s+[-*]\s?$/.test( currentLine ) ) {
```

Comment:
after playing around with composing the RegExps, I think I want to leave them alone for now. I don't see much gain in combining them and I don't see much gain in composing them when there are just two uses of them and those uses aren't identical and they aren't complicated.

thoughts @rodrigoi ?